applications/MeshingApplication/custom_processes/metrics_spr_error_process.cpp

```diff
@@ -0,0 +1,593 @@
+    BoundedMatrix<double, TDim + 1, TDim + 1> A = ZeroMatrix(TDim + 1,TDim + 1);
+    BoundedMatrix<double, TDim + 1, SigmaSize> b = ZeroMatrix(TDim + 1,SigmaSize);
+    BoundedMatrix<double, 1, TDim + 1> p_k;
+    
+    auto& neigh_elements = itPatchNode->GetValue(NEIGHBOUR_ELEMENTS);
+    for( WeakElementItType it_elem = neigh_elements.begin(); it_elem != neigh_elements.end(); ++it_elem) {
+        
+        it_elem->GetValueOnIntegrationPoints(variable_stress,stress_vector,mThisModelPart.GetProcessInfo());
+        it_elem->GetValueOnIntegrationPoints(variable_coordinates,coordinates_vector,mThisModelPart.GetProcessInfo());
+
+        KRATOS_INFO_IF("SPRMetricProcess", mEchoLevel > 3)
+        << "\tStress: " << stress_vector[0] << std::endl
+        << "\tx: " << coordinates_vector[0][0] << "\ty: " << coordinates_vector[0][1] << "\tz_coordinate: " << coordinates_vector[0][2] << std::endl;
+        
+        BoundedMatrix<double, 1, SigmaSize> sigma;
+        for(IndexType j = 0; j < SigmaSize; ++j)
+            sigma(0,j) = stress_vector[0][j];
+        p_k(0,0) = 1.0;
+        p_k(0,1) = coordinates_vector[0][0] - itPatchNode->X();
+        p_k(0,2) = coordinates_vector[0][1] - itPatchNode->Y();
+        if(TDim == 3)
+            p_k(0,3)=coordinates_vector[0][2] - itPatchNode->Z();
+        
+        noalias(A) += prod(trans(p_k), p_k);
+        noalias(b) += prod(trans(p_k), sigma);
+    }
+    
+    double det;
+    BoundedMatrix<double, TDim + 1, TDim + 1> invA = MathUtils<double>::InvertMatrix<TDim + 1>(A, det, -1.0); // We consider a negative tolerance in order to avoid error
+
+    KRATOS_INFO_IF("SPRMetricProcess", mEchoLevel > 3) << A << std::endl << invA << std::endl << det<< std::endl;
+
+    const double tolerance = std::numeric_limits<double>::epsilon();
+    if(det < tolerance){
+        KRATOS_WARNING_IF("SPRMetricProcess", mEchoLevel == 2) << A << std::endl;
+        for( IndexType i = 0; i < TDim + 1;i++){
+            for( IndexType j = 0; j < TDim + 1; j++)
+                A(i,j) += 0.001;
+        }
+        invA = MathUtils<double>::InvertMatrix<TDim + 1>(A,det);
+        KRATOS_WARNING_IF("SPRMetricProcess", mEchoLevel > 0) << "det: " << det << std::endl;
+    }
+
+    const BoundedMatrix<double, TDim + 1, SigmaSize> coeff = prod(invA, b);
+    
+    if(NeighbourSize > TDim) {
+        noalias(rSigmaRecovered) = row(coeff, 0);
+    } else {
+        p_k(0,1) = itNode->X() - itPatchNode->X();
+        p_k(0,2) = itNode->Y() - itPatchNode->Y();
+        if(TDim ==3)
+            p_k(0,3) = itNode->Z() - itPatchNode->Z();
+        const BoundedMatrix<double, 1, SigmaSize> sigma = prod(p_k,coeff);
+        noalias(rSigmaRecovered) = row(sigma, 0);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TDim>
+void SPRMetricProcess<TDim>::CalculatePatchContact(
```

Comment:
this function has over 200 lines, can you refactor?

Reply:
Difficult, everything is related